### ISSUE TITLE
Concat layers of different output sizes

### ISSUE THREAD

**Issue by lukemurray** [tiny-dnn/tiny-dnn#458]
Hi I'm trying to implement Squeezenet 1.1 (https://github.com/DeepScale/SqueezeNet/blob/master/SqueezeNet_v1.1/train_val.prototxt)

I can't seem to get the concat of the to (left, right) conv layers int he fire block as they are different sizes. One has a kernel size of 1 and the other has a kernel size of 3.

Not sure if I am doing something wrong or does tiny_dnn not support this action?

Here is a snippet of my code.

```
// fire block
convolutional_layer<relu> left(conv1.out_shape()[0].width_, conv1.out_shape()[0].height_, 1, 1, expand, padding::valid, true);
convolutional_layer<relu> right(conv1.out_shape()[0].width_, conv1.out_shape()[0].height_, 3, 3, expand, padding::valid, true);

concat_layer concat({left.out_shape()[0], right.out_shape()[0]}); // i've tried a few different things here
(left, right) << concat;
```

I'm coming from Keras/Tensorflow.

Thanks

**Comment by bhack** [comment]
Is it related to https://github.com/tiny-dnn/tiny-dnn/issues/221?

**Comment by lukemurray** [comment]
Sure we're both looking at squeeze net. I guess my question is, is it possible to concat different size inputs? It seems like it isn't. Is that functionality planned?

**Comment by bhack** [comment]
As I rembeber generally the mismatch size need to be limited on concact axis but actually we don't support axis specification in concat. We are also implementing  a new Tensor structure so I ask to @edgarriba and @nyanp if it makes sense to try to have this befor Tensor integration.

**Comment by lukemurray** [comment]
Ahh, yes I want to specify axis `3` but obviously can't. Would be awesome if this functionality was supported.

**Comment by bhack** [comment]
https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/layers/concat_layer.h#L34

**Comment by lukemurray** [comment]
Are you saying it is supported? I guess I'm also asking for an example. I've tried a few different ways on constructing it and I always get a runtime error.

**Comment by bhack** [comment]
I pointed you only on the behaviour description of the actual concact implementation

>concat N layers along depth

**Comment by Randl** [comment]
@lukemurray Citing from paper:

> So that the output activations from 1x1 and 3x3 filters have the same height and width, we
add a 1-pixel border of zero-padding in the input data to 3x3 filters of expand modules.

Try:

```cpp
// fire block
convolutional_layer<relu> left(conv1.out_shape()[0].width_, conv1.out_shape()[0].height_, 1, 1, expand, padding::valid, true);
convolutional_layer<relu> right(conv1.out_shape()[0].width_, conv1.out_shape()[0].height_, 3, 3, expand, padding::same, true);

concat_layer concat({left.out_shape()[0], right.out_shape()[0]}); 
(left, right) << concat;
```
